### DISCUSSION TITLE
Can't install Ruby with jemalloc on macOS Catalina 10.15.1

### DISCUSSION THREAD

**Discussion by ndbroadbent** [rbenv/ruby-build#1843] [Build failures]
Hello,

I just set up a new mac, and now `RUBY_CONFIGURE_OPTS=--with-jemalloc rbenv install 2.6.5` is failing with this error:

```
checking for utime.h... yes
checking gmp.h usability... no
checking gmp.h presence... no
checking for gmp.h... no
checking for library containing malloc_conf... no
checking jemalloc/jemalloc.h usability... no
checking jemalloc/jemalloc.h presence... no
checking for jemalloc/jemalloc.h... no
checking for jemalloc with JEMALLOC_MANGLE... no
configure: error: jemalloc requested but not found
```

I ran `brew install rbenv jemalloc` to install rbenv 1.1.2 and jemalloc 5.2.1.

How can I fix this issue? Thanks!

**Comment by mislav** [comment]
Hi thank you for reporting! I have no experience with jemalloc, and I haven't seen this error before. For me it works, but my mac is far from a pristine state.

Can you double check that jemalloc is linked into your `/usr/local/bin`? I.e. you can check whether running `jemalloc-config --bindir` works

**Comment by ndbroadbent** [comment]
Hello, here is the output:

```
$ jemalloc-config --bindir
/usr/local/Cellar/jemalloc/5.2.1/bin

$ ls -l /usr/local/bin/ | grep jemall
lrwxr-xr-x  1 ndbroadbent  admin        44 Dec  4 12:35 jemalloc-config -> ../Cellar/jemalloc/5.2.1/bin/jemalloc-config
lrwxr-xr-x  1 ndbroadbent  admin        40 Dec  4 12:35 jemalloc.sh -> ../Cellar/jemalloc/5.2.1/bin/jemalloc.sh
lrwxr-xr-x  1 ndbroadbent  admin        35 Dec  4 12:35 jeprof -> ../Cellar/jemalloc/5.2.1/bin/jeprof
```

Thanks for your help!

**Comment by mislav** [comment]
Well, I'm stumped. https://github.com/ruby/ruby/blob/v2_6_5/configure.ac#L1123-L1131

Not sure what is different in your environment vs mine.

**Comment by paulcsmith** [comment]
@ndbroadbent I occasionally run into this as well and end up having to uninstall homebrew and reinstall from scratch because I haven't figured out any other way to resolve it. Any tips?

**Comment by jaydorsey** [comment]
I'm seeing the same/similar behavior on Big Sur (m1) & Catalina (intel). I've tried a bunch of things to resolve this.

- I've tried `RUBY_CONFIGURE_OPTS="--with-jemalloc-dir=($brew --prefix jemalloc)"` (build succeeds, but i don't see jemalloc in [RbConfig::CONFIG['MAINLIBS']](https://bugs.ruby-lang.org/issues/15470))
- When I pass in `--with-jemalloc` (no directory) the build fails (same as OP)
- `jemalloc-config --bindir` works (returns a valid path/values)
- I have no jemalloc in `/opt/homebrew/bin` (but there is a `jemalloc.sh` symlink). that folder is exposed/available in my $path
- I'm using ruby-build via asdf-ruby for my version manager, but i've also tried directly with ruby-build

      export RUBY_CONFIGURE_OPTS="--with-jemalloc-dir=$(brew --prefix jemalloc) --with-readline-dir=($(brew --prefix readline)"
      ruby-build 3.0.0 ruby-3.0.0

- This was a pretty vanilla install on both machines (both new computers that I installed from scratch). Both running newer versions of homebrew. Force reinstalling jemalloc via homebrew doesn't resolve it
- I also completely re-installed homebrew from scratch after removing it

Everything looks normal to me, but it won't build with jemalloc on macOS. As far as I can tell, it hasn't for a while (I assumed it was, because it definitely did at one point). I went back and checked some older versions on another machine I had installed.

![image](https://user-images.githubusercontent.com/191564/109079479-fcb41b80-76cc-11eb-9964-fd90150c4e22.png)

Similar-ish output from my intel as well (latest versions of nearly everything)

I feel like this isn't even limited to jemalloc, because I see similar behavior with gmp (I have it, ruby-build can't detect it, but i see linux builds using it) but the cause might be the same. it's like nothing is getting picked up as a configure option

using zsh/asdf (not sure if those matter, also tried ruby-build)

happy to setup a time to pair/screenshare

Update: I followed some of the config docs for readline, with same/similar path settings for jemalloc:

```
export LDFLAGS="-L$HOMEBREW_PREFIX/opt/openssl@1.1/lib -L$HOMEBREW_PREFIX/opt/readline/lib -L$HOMEBREW_PREFIX/opt/jemalloc/lib"
export CPPFLAGS="-I$HOMEBREW_PREFIX/opt/openssl@1.1/include -I$HOMEBREW_PREFIX/opt/readline/include -I$HOMEBREW_PREFIX/opt/jemalloc/include"
```
With those changes, `RUBY_CONFIGURE_OPTS=--with-jemalloc asdf install ruby 3.0.0` build fine & `ruby -e "puts RbConfig::CONFIG['MAINLIBS']"` shows `-ljemalloc` as expected now

Also, I noticed some inconsistencies with the way configure/build flags behave so maybe I'm just doing this wrong now. `--with-jemalloc` works if i have the right *FLAGS, but `--with-jemalloc-dir` appears to do nothing. And it's the inverse for `gmp`; once i pass in `--with-gmp-dir` it works but `--with-gmp` doesn't

**Reply by jaydorsey** [reply]
FYI I recently saw this, maybe related (issues building gmp): https://bugs.ruby-lang.org/issues/20494

My original post noted that `--with-gmp` didn't work but I didn't investigate it beyond that (someone else did though)

**Comment by mostlyobvious** [comment]
> Update: I followed some of the config docs for readline, with same/similar path settings for jemalloc:


This worked for me (in fish shell):

```sh
set -gx HOMEBREW_PREFIX "/opt/homebrew";
set -gx CPPFLAGS "$CPPFLAGS -I $HOMEBREW_PREFIX/include"
set -gx LDFLAGS "$LDFLAGS -L $HOMEBREW_PREFIX/lib"
```

**Reply by heaven** [reply]
I've added this to my zshrc
```
# https://apple.stackexchange.com/questions/148901/why-does-my-brew-installation-not-work
eval $(/opt/homebrew/bin/brew shellenv)

# https://github.com/rbenv/ruby-build/issues/1388
export CPPFLAGS="$CPPFLAGS -I $HOMEBREW_PREFIX/include"
export LDFLAGS="$LDFLAGS -L $HOMEBREW_PREFIX/lib"
```

That helped a bit and it can find the headers but still complains that the library can't be found and fails.
```
checking gmp.h usability... yes
checking gmp.h presence... yes
checking for gmp.h... yes
checking for library containing __gmpz_init... no
checking jemalloc/jemalloc.h usability... yes
checking jemalloc/jemalloc.h presence... yes
checking for jemalloc/jemalloc.h... yes
checking for jemalloc library... no
configure: error: jemalloc requested but not found
```

**Reply by BiosBoy** [reply]
same issue here. Have you found a solution?

**Reply by ashwinv11** [reply]
This worked for me using the `asdf` version manager (which uses `ruby-build` under the hood) – https://github.com/asdf-vm/asdf-ruby/issues/117#issuecomment-1585820224

**Reply by pjmartorell** [reply]
I can confirm the above-mentioned flags worked with Zsh and Rbenv in my Apple M2 Pro with Sonoma 14.4.1:

```ruby
export CPPFLAGS="$CPPFLAGS -I $HOMEBREW_PREFIX/include"
export LDFLAGS="$LDFLAGS -L $HOMEBREW_PREFIX/lib"
RUBY_CONFIGURE_OPTS="--enable-yjit --with-jemalloc" rbenv install 3.3.3

ruby-build: using openssl@3 from homebrew
==> Downloading ruby-3.3.3.tar.gz...
-> curl -q -fL -o ruby-3.3.3.tar.gz https://cache.ruby-lang.org/pub/ruby/3.3/ruby-3.3.3.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 21.0M  100 21.0M    0     0  44.8M      0 --:--:-- --:--:-- --:--:-- 44.7M
==> Installing ruby-3.3.3...
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/3.3.3" --with-openssl-dir=/opt/homebrew/opt/openssl@3 --enable-shared --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+ --enable-yjit --with-jemalloc
-> make -j 10
-> make install
==> Installed ruby-3.3.3 to /Users/xxxxxx/.rbenv/versions/3.3.3
```
